EssentialsX Discord Features/Fixup Day 2

JRoy opened the pull request:
Summary of Changes;
* Added a advancement message type:
  ```yml
  message-types:
    ...
    # Achievement/Advancement messages sent when a player is awarded an achievement/advancement.
    advancement: primary
  ```
  ```yml
  messages:
    ...
    # This is the message sent to discord when a player is awarded an advancement.
    # The following placeholders can be used here:
    # - {username}: The name of the user who was awarded the advancement.
    # - {displayname}: The display name of the user who was awarded the advancement.
    # - {advancement}: The name of the advancement.
    # ... PlaceholderAPI placeholders are also supported here too!
    advancement: ":medal: {displayname} has completed the advancement **{advancement}**"
  ```
* Added option to show channels in console:
  ```yml
  chat:
    ...
    # A list of Discord channels which should be logged to console.
    # If you want to use the channels bellow, make sure to remove the '#' before the '-'.
    relay-to-console:
      #- primary
      #- chat
  ```
* Added the following config options to manage vanished users: 
  ```yml
  # Whether or not fake join and leave messages should be sent to discord when a player vanishes/un-vanishes in Minecraft.
  # Fake join/leave messages will be sent the same as real join and leave messages (and to the same channel).
  vanish-fake-join-leave: true
  
  # Whether or not messages from vanished players should be sent to discord.
  # This affects join, leave, death, and afk message types by default.
  vanish-hide-messages: true
  ```
* Added an option to show display name as webhook name:
  ```yml
  # Whether or not player messages should show their display-name/nickname as the bot name in Discord.
  # You must disable show-name for this option to work.
  show-displayname: false
  ```
* Changed the console timespamp placeholder to use native Discord timestamps.
* Fixed occasional exceptions with the console logger.
* Fixed join and quit messages not always showing when they should.
* Fixed numerous errors when EssentialsX Discord had invalid configs.
* Fixed the role color placeholder not working.
* Fixed death messages sometimes showing when they shouldn't.
* Fixed prefixes/suffixes sometimes showing hex format in discord.
* Fixed disabling one command causing all commands to be disabled.

JRoy committed:
Add fake vanish join/leave + exclude vanished messages

JRoy committed:
Ensure no JDA code is called during invalid shutdown

JRoy committed:
Fix disabling a single command causing all commands to be disabled

JRoy committed:
Add advancement message type

mdcfe commented on EssentialsDiscord/src/main/java/net/essentialsx/discord/JDADiscordService.java:
```diff
@@ -172,6 +184,17 @@ public void startup() throws LoginException, InterruptedException {
+                Bukkit.getPluginManager().registerEvents(new AchievementListenerProvider(), plugin);
+            }
+        } catch (final Throwable e) {
+            logger.log(Level.WARNING, "There was a problem with loading the advancement listener, you will not receive advancement notifications in discord.", e);
```
```suggestion
            logger.log(Level.WARNING, "Error while loading the achievement/advancement listener. You will not receive achievement/advancement notifications on Discord.", e);
```

mdcfe commented on EssentialsDiscord/src/main/java/net/essentialsx/discord/listeners/BukkitListener.java:
```diff
@@ -104,7 +107,7 @@ public void onChat(AsyncPlayerChatEvent event) {
     public void onJoin(AsyncUserDataLoadEvent event) {
         // Delay join to let nickname load
-        if (event.getJoinMessage() != null) {
+        if (event.getJoinMessage() != null && (!jda.getSettings().isVanishHideMessages() || !event.getUser().isHidden())) {
```
Maybe pull the new check out to a `boolean shouldShowJoinQuitMessage(User user)` method (or with a better name, maybe `shouldShowActivityMessage` so can be used later below)

Could maybe even put this in API - "by default" in the config comment implies other message types could respect this, and possibly worth exposing more easily to 3rd-party plugins, but it's up to you

mdcfe commented on EssentialsDiscord/src/main/java/net/essentialsx/discord/listeners/BukkitListener.java:
```diff
@@ -119,7 +122,7 @@ public void onJoin(AsyncUserDataLoadEvent event) {
     @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
     public void onQuit(PlayerQuitEvent event) {
-        if (event.getQuitMessage() != null) {
+        if (event.getQuitMessage() != null && (!jda.getSettings().isVanishHideMessages() || !jda.getPlugin().getEss().getUser(event.getPlayer()).isHidden())) {
```
Ditto

mdcfe commented on EssentialsDiscord/src/main/java/net/essentialsx/discord/listeners/BukkitListener.java:
```diff
@@ -132,8 +135,39 @@ public void onQuit(PlayerQuitEvent event) {
+        if (!jda.getSettings().isVanishFakeJoinLeave()) {
+            return;
+        }
+        if (event.getValue()) {
+            sendDiscordMessage(MessageType.DefaultTypes.LEAVE,
+                    MessageUtil.formatMessage(jda.getSettings().getQuitFormat(event.getAffected().getBase()),
+                            MessageUtil.sanitizeDiscordMarkdown(event.getAffected().getName()),
+                            MessageUtil.sanitizeDiscordMarkdown(event.getAffected().getDisplayName()),
+                            MessageUtil.sanitizeDiscordMarkdown(ChatColor.YELLOW + event.getAffected().getName() + " left the game")),
+                    false,
+                    jda.getSettings().isShowAvatar() ? AVATAR_URL.replace("{uuid}", event.getAffected().getBase().getUniqueId().toString()) : null,
+                    jda.getSettings().isShowName() ? event.getAffected().getName() : null,
+                    event.getAffected().getBase().getUniqueId());
+            return;
+        }
+        sendDiscordMessage(MessageType.DefaultTypes.JOIN,
+                MessageUtil.formatMessage(jda.getSettings().getJoinFormat(event.getAffected().getBase()),
+                        MessageUtil.sanitizeDiscordMarkdown(event.getAffected().getName()),
+                        MessageUtil.sanitizeDiscordMarkdown(event.getAffected().getDisplayName()),
+                        MessageUtil.sanitizeDiscordMarkdown(ChatColor.YELLOW + event.getAffected().getName() + " joined the game")),
+                false,
+                jda.getSettings().isShowAvatar() ? AVATAR_URL.replace("{uuid}", event.getAffected().getBase().getUniqueId().toString()) : null,
+                jda.getSettings().isShowName() ? event.getAffected().getName() : null,
+                event.getAffected().getBase().getUniqueId());
```
Pull out `get(Join|Quit)Format` and the default messages to deduplicate this

mdcfe commented on EssentialsDiscord/src/main/java/net/essentialsx/discord/listeners/BukkitListener.java:
```diff
@@ -132,8 +135,39 @@ public void onQuit(PlayerQuitEvent event) {
+
     @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
     public void onDeath(PlayerDeathEvent event) {
+        if (jda.getSettings().isVanishHideMessages() && jda.getPlugin().getEss().getUser(event.getEntity()).isHidden()) {
```
Ditto

mdcfe commented on EssentialsDiscord/src/main/java/net/essentialsx/discord/listeners/BukkitListener.java:
```diff
@@ -147,6 +181,10 @@ public void onDeath(PlayerDeathEvent event) {
 
     @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
     public void onAfk(AfkStatusChangeEvent event) {
+        if (jda.getSettings().isVanishHideMessages() && event.getAffected().isHidden()) {
```
Ditto

mdcfe commented on EssentialsDiscord/src/main/java/net/essentialsx/discord/listeners/BukkitListener.java:
```diff
@@ -164,8 +202,28 @@ public void onAfk(AfkStatusChangeEvent event) {
 
+    @EventHandler(priority = EventPriority.MONITOR)
+    public void onAdvancement(AbstractAchievementEvent event) {
+        if (jda.getSettings().isVanishHideMessages() && jda.getPlugin().getEss().getUser(event.getPlayer()).isHidden()) {
```
Ditto

mdcfe commented on EssentialsDiscord/src/main/java/net/essentialsx/discord/listeners/BukkitListener.java:
```diff
@@ -164,8 +202,28 @@ public void onAfk(AfkStatusChangeEvent event) {
+
     @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
     public void onKick(PlayerKickEvent event) {
+        if (jda.getSettings().isVanishHideMessages() && jda.getPlugin().getEss().getUser(event.getPlayer()).isHidden()) {
```
![image](https://user-images.githubusercontent.com/1917406/124495076-8bd5ec00-ddaf-11eb-9042-47636ce5ecab.png)

mdcfe commented on EssentialsDiscord/src/main/resources/config.yml:
```diff
@@ -110,6 +110,8 @@ message-types:
   death: primary
   # AFK status change messages sent when a player's AFK status changes.
   afk: primary
+  # Achievement/Advancement messages sent when a player is awarded an achievement/advancement.
```
```suggestion
  # Achievement/advancement messages sent when a player is awarded an achievement/advancement.
```

mdcfe commented on EssentialsDiscord/src/main/resources/config.yml:
```diff
@@ -124,6 +126,14 @@ show-avatar: false
 # Whether or not player messages should show their name as the bot name in Discord.
 show-name: false
 
+# Whether or not fake join and leave messages should be sent to discord when a player vanishes/un-vanishes in Minecraft.
+# Fake join/leave messages will be sent the same as real join and leave messages (and to the same channel).
```
```suggestion
# Whether or not fake join and leave messages should be sent to Discord when a player toggles vanish in Minecraft.
# Fake join/leave messages will be sent the same as real join and leave messages (and to the same channel).
```

mdcfe commented on EssentialsDiscord/src/main/resources/config.yml:
```diff
@@ -124,6 +126,14 @@ show-avatar: false
+# Fake join/leave messages will be sent the same as real join and leave messages (and to the same channel).
+vanish-fake-join-leave: true
+
+# Whether or not messages from vanished players should be sent to discord.
```
```suggestion
# Whether or not messages from vanished players should be sent to Discord.
```

mdcfe commented on EssentialsDiscord/src/main/resources/config.yml:
```diff
@@ -299,6 +305,17 @@ messages:
   # - {displayname}: The display name of the user who is no longer afk
   # ... PlaceholderAPI placeholders are also supported here too!
   un-afk: ":keyboard: {displayname} is no longer AFK!"
+  # This is the message sent to discord when a player is awarded an advancement.
```
```suggestion
  # This is the message sent to Discord when a player is awarded an advancement.
```

mdcfe commented on EssentialsDiscord/src/main/resources/config.yml:
```diff
@@ -299,6 +305,17 @@ messages:
+  # - {advancement}: The name of the advancement.
+  # ... PlaceholderAPI placeholders are also supported here too!
+  advancement: ":medal: {displayname} has completed the advancement **{advancement}**!"
+  # This is the message sent to discord when the server starts.
```
```suggestion
  # This is the message sent to Discord when the server starts.
```

I realise this isn't new but it slipped through previously

Also, why move these? Will introduce confusion for people updating configs, and I don't see a reason to move these all the way down here?

JRoy replied:
Mistake from before, I want player-specific (and thus PAPI supporting) message types on the top, others on bottom. Not too worried about it, we haven't done a full release yet.

mdcfe commented on EssentialsDiscord/src/main/resources/config.yml:
```diff
@@ -299,6 +305,17 @@ messages:
+  advancement: ":medal: {displayname} has completed the advancement **{advancement}**!"
+  # This is the message sent to discord when the server starts.
+  server-start: ":white_check_mark: The server has started!"
+  # This is the message sent to discord when the server stops.
```
```suggestion
  # This is the message sent to Discord when the server stops.
```
![jroyditto](https://user-images.githubusercontent.com/1917406/124496070-ddcb4180-ddb0-11eb-8039-341d3080766e.png)

JRoy committed:
Fix prefixes/suffixes not stripping hex format

JRoy committed:
Fix improper bitwise logic in DiscordUtil#getRoleColorFormat

JRoy committed:
Improve death message validation

Now takes gamerules in to account and will not send blank or null death messages.

JRoy committed:
Fix some minor grammer issues in config

JRoy committed:
Add option to log channels to console

mdcfe commented on EssentialsDiscord/src/main/java/net/essentialsx/discord/listeners/BukkitListener.java:
```diff
@@ -104,36 +110,74 @@ public void onChat(AsyncPlayerChatEvent event) {
+        sendJoinMessage(event.getAffected().getBase(), ChatColor.YELLOW + event.getAffected().getName() + " joined the game");
+    }
+
+    public void sendJoinMessage(final Player player, final String joinMessage) {
+        sendDiscordMessage(MessageType.DefaultTypes.JOIN,
+                MessageUtil.formatMessage(jda.getSettings().getJoinFormat(player),
+                        MessageUtil.sanitizeDiscordMarkdown(player.getName()),
+                        MessageUtil.sanitizeDiscordMarkdown(player.getDisplayName()),
+                        MessageUtil.sanitizeDiscordMarkdown(joinMessage),
+                false,
+                jda.getSettings().isShowAvatar() ? AVATAR_URL.replace("{uuid}", player.getUniqueId().toString()) : null,
+                jda.getSettings().isShowName() ? player.getName() : null,
+                        player.getUniqueId()));
+    }
+
+    public void sendQuitMessage(final Player player, final String quitMessage) {
+        sendDiscordMessage(MessageType.DefaultTypes.LEAVE,
+                MessageUtil.formatMessage(jda.getSettings().getQuitFormat(player),
+                        MessageUtil.sanitizeDiscordMarkdown(player.getName()),
+                        MessageUtil.sanitizeDiscordMarkdown(player.getDisplayName()),
+                        MessageUtil.sanitizeDiscordMarkdown(quitMessage)),
+                false,
+                jda.getSettings().isShowAvatar() ? AVATAR_URL.replace("{uuid}", player.getUniqueId().toString()) : null,
+                jda.getSettings().isShowName() ? player.getName() : null,
+                player.getUniqueId());
+    }
```
These don't need to be separate methods, just pass in the type and format?

mdcfe commented on EssentialsDiscord/src/main/resources/config.yml:
```diff
@@ -61,6 +61,11 @@ chat:
   # You shouldn't need to enable this unless you're not seeing all chat messages go through to Discord.
   show-all-chat: false
+  # A list of Discord channels which should be logged to console.
+  # If you want to use the channels bellow, make sure to remove the '#' before the '-'.
```
```suggestion
  # If you want to use the channels below, remember to uncomment the line by removing the '#' before the '-'.
```

JRoy replied:
lmfao

mdcfe commented:
Also, should update the readme and wiki to document the new features in this PR

Chew committed:
Use native Discord timestamps for console timestamps

JRoy committed:
Fix less strict length check on messageQueue add rather than iteration

JRoy committed:
Remove join/quit message null checks

If you don't want to see them, don't configure them.

mdcfe commented on EssentialsDiscord/src/main/resources/config.yml:
```diff
@@ -51,8 +51,8 @@ chat:
   filter-newlines: true
-  # A regex pattern which will not send matching messages through to Discord.
-  # By default, this will ignore messages starting with '!' and '?'.
+  # A regex pattern which will not send matching messages through from Discord.
```
```suggestion
  # A regex pattern which will ignore matching incoming messages from Discord.
```

JRoy replied:
done lulw

JRoy committed:
Add show-displayname option

mdcfe commented on EssentialsDiscord/src/main/java/net/essentialsx/discord/DiscordSettings.java:
```diff
@@ -298,6 +326,11 @@ private MessageFormat generateMessageFormat(String content, String defaultStr, b
     @Override
     public void reloadConfig() {
+        if (plugin.isInvalidStartup()) {
+            plugin.getLogger().warning("Tried to reload EssentialsX Discord config while the plugin is in an invalid state!");
```
Might want to show a message about "if you've modified your config, restart your server"

mdcfe commented on EssentialsDiscord/src/main/java/net/essentialsx/discord/JDADiscordService.java:
```diff
@@ -172,6 +184,17 @@ public void startup() throws LoginException, InterruptedException {
         Bukkit.getPluginManager().registerEvents(new BukkitListener(this), plugin);
+
+        try {
+            if (VersionUtil.getServerBukkitVersion().isHigherThan(VersionUtil.v1_12_2_R01)) {
```
Advancements are 1.12, not 1.12.2. (There's not any reason for people to be on 1.12/1.12.1 but still better to use the correct versions if we're not doing reflective feature detection any more)
